### PR TITLE
[WIP] Feature Argument Hints

### PR DESCRIPTION
**Pull request by binaryanthems**

### CODE REVIEW

File: argmax/Cargo.toml [hunk 1]
```diff
@@ -16,5 +16,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline argmax/Cargo.toml:19]: Please revert this change. Enabling the `metadata` feature by default will break any Runes that use this proc-block because of the environment a Rune is compiled in (`no_std`, etc.).

File: audio_float_conversion/Cargo.toml [hunk 2]
```diff
@@ -18,5 +18,5 @@ pretty_assertions = "0.7.2"
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline audio_float_conversion/Cargo.toml:21]:
```suggestion
default = []
```

File: binary_classification/Cargo.toml [hunk 3]
```diff
@@ -15,5 +15,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline binary_classification/Cargo.toml:18]:
```suggestion
default = []
```

File: fft/Cargo.toml [hunk 4]
```diff
@@ -26,6 +26,6 @@ pretty_assertions = "0.7.2"
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline fft/Cargo.toml:29]:
```suggestion
default = []
```

File: image-normalization/Cargo.toml [hunk 5]
```diff
@@ -16,5 +16,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline image-normalization/Cargo.toml:19]:
```suggestion
default = []
```

File: label/Cargo.toml [hunk 6]
```diff
@@ -17,5 +17,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline label/Cargo.toml:20]:
```suggestion
default = []
```

File: modulo/Cargo.toml [hunk 7]
```diff
@@ -17,6 +17,6 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline modulo/Cargo.toml:20]:
```suggestion
default = []
```

File: most_confident_indices/Cargo.toml [hunk 8]
```diff
@@ -15,5 +15,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline most_confident_indices/Cargo.toml:18]:
```suggestion
default = []
```

File: noise-filtering/Cargo.toml [hunk 9]
```diff
@@ -17,5 +17,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline noise-filtering/Cargo.toml:20]:
```suggestion
default = []
```

File: normalize/Cargo.toml [hunk 10]
```diff
@@ -16,5 +16,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline normalize/Cargo.toml:19]:
```suggestion
default = []
```

File: object_filter/Cargo.toml [hunk 11]
```diff
@@ -16,5 +16,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline object_filter/Cargo.toml:19]:
```suggestion
default = []
```

File: parse/Cargo.toml [hunk 12]
```diff
@@ -15,5 +15,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline parse/Cargo.toml:18]:
```suggestion
default = []
```

File: segment_output/Cargo.toml [hunk 13]
```diff
@@ -15,5 +15,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline segment_output/Cargo.toml:18]:
```suggestion
default = []
```

File: softmax/Cargo.toml [hunk 14]
```diff
@@ -16,6 +16,6 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline softmax/Cargo.toml:19]:
```suggestion
default = []
```

File: text_extractor/Cargo.toml [hunk 15]
```diff
@@ -15,6 +15,6 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline text_extractor/Cargo.toml:18]:
```suggestion
default = []
```

File: tokenizers/Cargo.toml [hunk 16]
```diff
@@ -18,5 +18,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline tokenizers/Cargo.toml:21]:
```suggestion
default = []
```

File: utf8_decode/Cargo.toml [hunk 17]
```diff
@@ -15,5 +15,5 @@ wit-bindgen-rust = { git = "https://github.com/bytecodealliance/wit-bindgen", op
 crate-type = ["rlib", "cdylib"]
 
 [features]
-default = []
+default = ["metadata"]
```
Comment (Michael-F-Bryan) [inline utf8_decode/Cargo.toml:18]:
```suggestion
default = []
```

File: fft/src/lib.rs [hunk 18]
```diff
@@ -149,16 +151,24 @@ pub mod metadata {
...
             bins.set_type_hint(TypeHint::Integer);
+            let bins_range = interpret_as_number_in_range("1", "1000");
+            bins.add_hint(&bins_range);
             metadata.add_argument(&bins);
 
             let window_overlap = ArgumentMetadata::new("window_overlap");
             window_overlap.set_default_value("0.6666667");
             window_overlap.set_type_hint(TypeHint::Float);
+            let window_overlap_range =
+                interpret_as_number_in_range("0.0000000", "1.0000000");
```
Comment (Michael-F-Bryan) [inline fft/src/lib.rs:170]: 7 zeroes is a bit overkill, don't you think?

```suggestion
                interpret_as_number_in_range("0.0", "1.0");
```

File: xtask/src/metadata.rs [hunk 19]
```diff
@@ -414,6 +409,54 @@ impl runtime_v1::RuntimeV1 for Runtime {
...
+    }
+
+    fn drop_tensor_metadata(&mut self, state: Self::TensorMetadata) {
+        drop(state);
+    }
+
+    fn drop_argument_hint(&mut self, state: Self::ArgumentHint) {
+        drop(state);
+    }
+
```
Comment (Michael-F-Bryan) [inline xtask/src/metadata.rs:432]: I don't think we need to explicitly `drop()` the `state` variables here. 

Values are automatically dropped whenever they go out of scope, and I'm pretty sure we get a default implementation for each of the `drop_XXX()` methods that does nothing except consume the `state`.

```suggestion
```

**Review by Michael-F-Bryan — CHANGES_REQUESTED** [review]
You'll need to revert the `default = ["metadata"]` changes to `Cargo.toml` before we can merge, but other than that I'm happy with the PR.

The `metadata` feature is an interim measure so we can start using WIT files the extracting metadata from proc-blocks while the actual computation stuff is done using the previous Rune-runtime interface. 

However, enabling it by default is problematic because it means we won't be able to load a Rune containing that proc-block. The proc-block will ask the runtime for the functions defined in our `*.wit` file, but the Rust, JavaScript, and C++ runtimes don't provide those functions yet.

Once we've rehashed the way proc-blocks are implemented, and our runtime libraries, I plan to remove the `metadata` feature flag entirely.

File: argmax/Cargo.toml [hunk 20]
```diff
@@ -8,6 +8,7 @@ repository = "https://github.com/hotg-ai/proc-blocks"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+wasm-bindgen = "0.2"
```
Comment (binaryanthems) [inline argmax/Cargo.toml:11]: Remove this line.

File: modulo/src/lib.rs [hunk 21]
```diff
@@ -66,6 +70,9 @@ pub mod metadata {
             let modulus = ArgumentMetadata::new("modulus");
             modulus.set_type_hint(TypeHint::Float);
             modulus.set_default_value("0");
+            let modulus_range =
+                interpret_as_number_in_range("0", "MAX_FLOAT64"); //Yes 65526 not 170 followed by another 306 zeros then a decimal point
```
Comment (binaryanthems) [inline modulo/src/lib.rs:74]: `Infinity`
```suggestion
                interpret_as_number_in_range("0", "Infinity");
```

File: argmax/Cargo.toml (hunk 20 above)
Comment (binaryanthems) [reply argmax/Cargo.toml:11]:
```suggestion
```
